tests/autorestart_test/test_container_autorestart.py

```diff
@@ -0,0 +1,340 @@
+"""
+Test the auto-restart feature of containers
+"""
+import logging
+import time
+from collections import defaultdict
+
+import pytest
+
+from tests.common.utilities import wait_until
+from tests.common.helpers.assertions import pytest_assert
+
+logger = logging.getLogger(__name__)
+
+pytestmark = [
+    pytest.mark.topology('any')
+]
+
+CONTAINER_CHECK_INTERVAL_SECS = 1
+CONTAINER_STOP_THRESHOLD_SECS = 30
+CONTAINER_RESTART_THRESHOLD_SECS = 180
+
+CMD_FEATURE = "show feature"
+CMD_CONTAINER_FEATURE_AUTORESTART = "show container feature autorestart"
+
+
+def get_critical_group_and_process_list(duthost, container_name):
+    """
+    @summary: Get critical group and process lists by parsing the
+              critical_processes file in the specified container
+    @return: Two lists which include the critical process and critical groups respectively
+    """
+    critical_group_list = []
+    critical_process_list = []
+
+    file_content = duthost.shell("docker exec {} bash -c '[ -f /etc/supervisor/critical_processes ] \
+            && cat /etc/supervisor/critical_processes'".format(container_name), module_ignore_errors=True)
+    for line in file_content["stdout_lines"]:
+        line_info = line.strip('\n').split(':')
+        if len(line_info) != 2:
+            pytest.fail("Syntax of the line '{}' in critical_processes file is incorrect.".format(line))
+
+        identifier_key = line_info[0].strip()
+        identifier_value = line_info[1].strip()
+        if identifier_key == "group" and identifier_value:
+            critical_group_list.append(identifier_value)
+        elif identifier_key == "program" and identifier_value:
+            critical_process_list.append(identifier_value)
+        else:
+            pytest.fail("Syntax of the line '{}' in critical_processes file is incorrect.".format(line))
+
+    return critical_group_list, critical_process_list
```

Comment:
What is the difference between this function and https://github.com/Azure/sonic-mgmt/blob/695a9a53b75c90878c8b8de022479a53af1240a0/tests/common/devices.py#L371? can you merge this into a common place?

This can come as another PR.

Reply:
The difference between these two functions is `all_critical_process_status(...)` will call another function `critical_process_status(...)` which first get all the critical process list and then retrieve their running status for all containers in SONiC, `get_critical_group_and_process_list(...)` will only get all critical process list. 

Can we break the function `critical_process_status(...)` in `common/devices.py ` into two sub-functions? One function is used to get all critical process list of a container and another function can be created to retrieve the running status of these critical processes.

Reply:
@yozhao101: Yes. I think splitting the function into two sub-functions should be fine.

Reply:
I will do this by creating a separate PR.